File: Word-Pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> sv;
        stringstream ss(s);
        string w;
        while(ss>>w){
            sv.push_back(w);
        }
        if(pattern.size()!=sv.size())return false;
        unordered_map<char,string> maps;
        for(int i=0;i<pattern.size();i++){
            if(maps.find(pattern[i])==maps.end())maps[pattern[i]]=sv[i];
            else if(maps[pattern[i]]!=sv[i])return false;
        }
        unordered_map<string,char> nmaps;
        for(int i=0;i<sv.size();i++){
            if(nmaps.find(sv[i])==nmaps.end())nmaps[sv[i]]=pattern[i];
            else if(nmaps[sv[i]]!=pattern[i])return false;
        }
        return true;
    }
};

File: Word-Pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        List<String> sv = new ArrayList<>();
        Scanner ss = new Scanner(s);
        while(ss.hasNext()){
            String w = ss.next();
            sv.add(w);
        }
        ss.close();

        if(pattern.length() != sv.size()) return false;

        HashMap<Character, String> maps = new HashMap<>();
        for(int i = 0; i < pattern.length(); i++){
            if(!maps.containsKey(pattern.charAt(i))) {
                maps.put(pattern.charAt(i), sv.get(i));
            } else if(!maps.get(pattern.charAt(i)).equals(sv.get(i))) {
                return false;
            }
        }

        HashMap<String, Character> nmaps = new HashMap<>();
        for(int i = 0; i < sv.size(); i++){
            if(!nmaps.containsKey(sv.get(i))) {
                nmaps.put(sv.get(i), pattern.charAt(i));
            } else if(!nmaps.get(sv.get(i)).equals(pattern.charAt(i))) {
                return false;
            }
        }
        return true;
    }
}